types/utils.d.ts

```diff
@@ -0,0 +1,79 @@
+import { ActionContext } from './index'
+
+/*
+ * Type level utilities to annotate types of getters/actions/mutations (module assets).
+ * To use the below helpers, the user should declare corresponding assets type at first.
+ *
+ * A getters type should be an object that the keys indicate getter names
+ * and its corresponding values indicate return types of the getter.
+ *
+ * Actions type and mutations type should be an object that the keys indicate
+ * action/mutation names as same as the getters type.
+ * Its values should be declared as payload types of the actions/mutation.
+ *
+ * After declare the above types, the user put them on the generic parameters
+ * of the utility types. Then the real assets object must follow the passed types
+ * and type inference will work.
+ *
+ * The declared types will be used on mapXXX helpers to safely use module assets
+ * by annotating its types.
+ */
+
+/**
+ * Infer getters object type from passed generic types.
+ * `Getters` is an object type that the keys indicate getter names and
+ * its corresponding values are return types of the getters.
+ * `State` is a module state type which is accessible in the getters.
+ * `ExtraGetters` is like `Getters` type but will be not defined in the infered getters object.
+ * `RootState` and `RootGetters` are the root module's state and getters type.
+ */
+export type DefineGetters<
```

Comment:
What about a one-for-all option? like : 
```ts

export type DefineStore<
  Actions,
  State,
  Getters,
  Mutations,
  ExtraActions = {},
  RootState = {},
  RootGetters = {},
  RootMutations = {},
  RootActions = {}
> = {
  // ....
}
```

Reply:
Yes, it looks better since we always combine all assets for module 😄 
I will rewrite the utility types.